streams/src/main/java/org/apache/kafka/streams/state/internals/BoundsCheckingWindowStore.java

```diff
@@ -0,0 +1,147 @@
+        delegate.put(key, value);
+    }
+
+    @Override
+    public void put(final K key, final V value, final long timestamp) {
+        final long expiryTime = getExpiryTime();
+        if (timestamp >= expiryTime) {
+            delegate.put(key, value, timestamp);
+        } else {
+            LOG.warn("Dropping put for expired timestamp: key=[{}] timestamp=[{}] expiration=[{}]", key, timestamp, expiryTime);
+        }
+    }
+
+    @Override
+    public String name() {
+        return delegate.name();
+    }
+
+    @Override
+    public void init(final ProcessorContext context, final StateStore root) {
+        this.context = context;
+        delegate.init(context, root);
+    }
+
+    @Override
+    public void flush() {
+        delegate.flush();
+    }
+
+    @Override
+    public void close() {
+        delegate.close();
+    }
+
+    @Override
+    public boolean persistent() {
+        return delegate.persistent();
+    }
+
+    @Override
+    public boolean isOpen() {
+        return delegate.isOpen();
+    }
+
+    @Override
+    public V fetch(final K key, final long time) {
+        if (time >= getExpiryTime()) {
+            return delegate.fetch(key, time);
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public WindowStoreIterator<V> fetch(final K key, final long timeFrom, final long timeTo) {
+        final long newerOfTimestampOrExpiryTime = getNewerOfTimestampOrExpiryTime(timeFrom);
+
+        if (timeTo < newerOfTimestampOrExpiryTime) {
+            return KeyValueIterators.emptyWindowStoreIterator();
+        } else {
+            return delegate.fetch(key, newerOfTimestampOrExpiryTime, timeTo);
+        }
+    }
+
+    @Override
+    public KeyValueIterator<Windowed<K>, V> fetch(final K from, final K to, final long timeFrom, final long timeTo) {
+        final long newerOfTimestampOrExpiryTime = getNewerOfTimestampOrExpiryTime(timeFrom);
+
+        if (timeTo < newerOfTimestampOrExpiryTime) {
+            return KeyValueIterators.emptyIterator();
+        } else {
+            return delegate.fetch(from, to, newerOfTimestampOrExpiryTime, timeTo);
+        }
+    }
+
+    @Override
+    public KeyValueIterator<Windowed<K>, V> all() {
+        return fetchAll(0, Long.MAX_VALUE);
+    }
+
+    @Override
+    public KeyValueIterator<Windowed<K>, V> fetchAll(final long timeFrom, final long timeTo) {
+        final long newerOfTimestampOrExpiryTime = getNewerOfTimestampOrExpiryTime(timeFrom);
+
+        if (timeTo < newerOfTimestampOrExpiryTime) {
+            return KeyValueIterators.emptyIterator();
+        } else {
+            return delegate.fetchAll(newerOfTimestampOrExpiryTime, timeTo);
+        }
+    }
+
+
+    private long getNewerOfTimestampOrExpiryTime(final long timestamp) {
+        return Math.max(timestamp, getExpiryTime());
+    }
+
+    private long getExpiryTime() {
+        if (InternalProcessorContext.class.isAssignableFrom(context.getClass())) {
```

Comment:
How could this be false?

Reply:
MockProcessorContext in particular doesn't implement InternalProcsesorContext.